controllers/istio_controller_test.go

```diff
@@ -59,21 +64,416 @@ var _ = Describe("Istio Controller", func() {
+				istioInstallation: &istioInstallationReconciliationMock{
+					err: described_errors.NewDescribedError(errors.New("istio test error"), "test error description"),
+				},
+				proxySidecars:          &proxySidecarsReconciliationMock{},
+				log:                    logr.Discard(),
+				statusHandler:          status.NewDefaultStatusHandler(),
+				reconciliationInterval: testReconciliationInterval,
+			}
+
+			// when
+			result, err := sut.Reconcile(context.Background(), reconcile.Request{NamespacedName: types.NamespacedName{Namespace: testNamespace, Name: istioCrName}})
+
+			// then
+			Expect(err).Should(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring("istio test error"))
+			Expect(result).Should(Equal(reconcile.Result{}))
+
+			Expect(fakeClient.Get(context.TODO(), client.ObjectKeyFromObject(istioCR), istioCR)).Should(Succeed())
+			Expect(istioCR.Status.State).Should(Equal(operatorv1alpha1.Error))
+			Expect(istioCR.Status.Description).To(ContainSubstring("test error description"))
+		})
+
+		It("Should set error status and return an error when proxy sidecar reconciliation failed", func() {
+			// given
+			istioCR := &operatorv1alpha1.Istio{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      istioCrName,
+					Namespace: testNamespace,
+					Finalizers: []string{
+						"istios.operator.kyma-project.io/istio-installation",
+					},
+				},
+			}
+
+			fakeClient := createFakeClient(istioCR)
+
+			sut := &IstioReconciler{
+				Client:            fakeClient,
+				Scheme:            getTestScheme(),
+				istioInstallation: &istioInstallationReconciliationMock{},
+				proxySidecars: &proxySidecarsReconciliationMock{
+					err: errors.New("sidecar test error"),
+				},
+				log:                    logr.Discard(),
+				statusHandler:          status.NewDefaultStatusHandler(),
+				reconciliationInterval: testReconciliationInterval,
+			}
+
+			// when
+			result, err := sut.Reconcile(context.Background(), reconcile.Request{NamespacedName: types.NamespacedName{Namespace: testNamespace, Name: istioCrName}})
+
+			// then
+			Expect(err).Should(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring("sidecar test error"))
+			Expect(result).Should(Equal(reconcile.Result{}))
+
+			Expect(fakeClient.Get(context.TODO(), client.ObjectKeyFromObject(istioCR), istioCR)).Should(Succeed())
+			Expect(istioCR.Status.State).Should(Equal(operatorv1alpha1.Error))
+			Expect(istioCR.Status.Description).To(ContainSubstring("Error occurred during reconciliation of Istio Sidecars"))
 		})
 	})
 })
 
 func createFakeClient(objects ...client.Object) client.Client {
-	err := operatorv1alpha1.AddToScheme(scheme.Scheme)
-	Expect(err).ShouldNot(HaveOccurred())
+	return fake.NewClientBuilder().WithScheme(getTestScheme()).WithObjects(objects...).Build()
+}
+
+func getTestScheme() *runtime.Scheme {
+	Expect(operatorv1alpha1.AddToScheme(scheme.Scheme)).Should(Succeed())
+
+	return scheme.Scheme
+}
+
+type istioInstallationReconciliationMock struct {
```

Comment:
did we consider using a mock framework for these generations, e.g. counterfeit, mockery, gomock etc ...

Reply:
I think we were using some of those in the reconciler and I remember that people didn't like it. 
I'm also not a big fan of this mock implementations I did here, but also most Go mocking frameworks I saw are just so cumbersome with all that generated code ... I'll check how we can replace this.

Reply:
After discussion we agreed to keep it as it is for now and see how things develop. If it gets more complex we should replace it.